common/achievement/create.ts

```diff
@@ -0,0 +1,96 @@
+                    },
+                };
+            }),
+        },
+        select: { id: true, userId: true, context: true, template: true, achievedAt: true, recordValue: true },
+    });
+    return userAchievement;
+}
+
+async function getOrCreateUserAchievement<ID extends ActionID>(template: Achievement_template, userId: string, context?: SpecificNotificationContext<ID>) {
+    const isGlobal =
+        template.templateFor === achievement_template_for_enum.Global ||
+        template.templateFor === achievement_template_for_enum.Global_Courses ||
+        template.templateFor === achievement_template_for_enum.Global_Matches;
+    const existingUserAchievement = await findUserAchievement(template.id, userId, !isGlobal && context);
+    if (!existingUserAchievement) {
+        return await createAchievement(template, userId, context);
+    }
+    return existingUserAchievement;
+}
+
+async function createAchievement<ID extends ActionID>(currentTemplate: Achievement_template, userId: string, context: SpecificNotificationContext<ID>) {
+    const templatesByGroup = await getAchievementTemplates(TemplateSelectEnum.BY_GROUP);
+    const keys = Object.keys(context);
+    const userAchievementsByGroup = await prisma.user_achievement.findMany({
+        where: {
+            template: {
+                group: currentTemplate.group,
+            },
+            userId,
+            AND: keys.map((key) => {
+                return {
+                    context: {
+                        path: key,
+                        equals: context[key],
+                    },
+                };
+            }),
+        },
+        orderBy: { template: { groupOrder: 'asc' } },
+    });
+
+    const nextStepIndex = userAchievementsByGroup.length > 0 ? userAchievementsByGroup.findIndex((e) => e.groupOrder === currentTemplate.groupOrder) + 1 : 0;
+
+    const templatesForGroup = templatesByGroup.get(currentTemplate.group);
+    if (templatesForGroup && templatesForGroup.length > nextStepIndex) {
+        const createdUserAchievement = await createNextUserAchievement(templatesForGroup, nextStepIndex, userId, context);
+        return createdUserAchievement;
+    }
+}
+
+async function createNextUserAchievement<ID extends ActionID>(
+    templatesForGroup: Achievement_template[],
+    nextStepIndex: number,
+    userId: string,
+    context: SpecificNotificationContext<ID>
+) {
+    const nextStepTemplate = templatesForGroup[nextStepIndex];
+    // Here a user template is created for the next template in the group. This is done to always have the data availible for the next step.
+    // This could mean to, for example, have the name of a match partner that is not yet availible due to a unfinished matching process.
+    if (nextStepTemplate && nextStepTemplate.isActive) {
+        const createdUserAchievement = await prisma.user_achievement.create({
+            data: {
+                userId: userId,
+                group: nextStepTemplate.group,
```

Comment:
How do we distinguish two match groups? Do we need to use handlebars for the group name as well?

Reply:
currently the match id is saved in the context, so it does not need to be included in the group name. Currently the group is just there to distinguish all achievements that belong to one tiered list / sequence.

Reply:
In this case we have to make sure that it's taking into account everywhere.
https://github.com/corona-school/backend/blob/c42ff66a901b889b5a443b659dcaf4f6653b1ad5/common/achievement/get.ts#L16-L22

If I understand the code correctly, it would stack user_achievements of the same group, but different matches together.

Reply:
you are right. This function was updated to meet the requirements:

    const userAchievementGroups: { [groupRelation: string]: User_achievement[] } = {};
    userAchievements.forEach((ua) => {
        const relation = ua.context['relation'] || null;
        const key = relation ? `${ua.template.group}/${relation}` : ua.template.group;
        if (!userAchievementGroups[key]) {
            userAchievementGroups[key] = [];
        }
        userAchievementGroups[key].push(ua);
    });